plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts

```diff
@@ -0,0 +1,261 @@
+
+    const outbound: Promise<void>[] = [];
+    await Promise.all(
+      entityRefs.map(async entityRef => {
+        const compoundEntityRef = parseEntityRef(entityRef);
+        // skip users as they are sent direct messages, but allow all other entity kids
+        // to have a channel id annotation.
+        if (compoundEntityRef.kind === 'user') {
+          return;
+        }
+
+        let destination;
+        try {
+          destination = await this.getChannelFromEntity(entityRef);
+        } catch (error) {
+          this.logger.error(
+            `Failed to get Slack channel for entity: ${error.message}`,
+          );
+          return;
+        }
+
+        if (!destination) {
+          this.logger.debug(`No Slack channel found for entity: ${entityRef}`);
+          return;
+        }
+
+        const payload = JSON.stringify({
+          ...options.payload,
+          destination,
+        });
+
+        this.logger.debug(`Sending Slack channel notification: ${payload}`);
+        outbound.push(
+          sendSlackNotification(this.channelWebhookUrl!, payload).catch(
+            error => {
+              this.logger.error(
+                `Failed to send Slack channel notification: ${error.message}`,
+              );
+            },
+          ),
+        );
+      }),
+    );
+
+    // this is a bit fire and forget, but I don't think we want to block further processing.
+    await Promise.all(outbound);
+
+    if (outbound.length > 0) {
+      // If we sent team notifications, we don't want to send individual ones after resolving the group
+      // This is how we prevent the other methods from sending individual notifications.
+      // Given the current logic, if a Group has no slack channel id annotation, we'll let it get expanded to individuals.
+      options.recipients.entityRef = [];
+    }
+    return options;
+  }
+
+  async postProcess(
+    notification: Notification,
+    options: NotificationSendOptions,
+  ): Promise<void> {
+    if (
+      !this.directWebhookUrl ||
+      options.recipients.type === 'broadcast' ||
+      !notification.user
+    ) {
+      return;
+    }
+
+    const destination = await this.getEmailFromUser(notification.user);
+    if (!destination) {
+      this.logger.debug(`No email found for user entity: ${notification.user}`);
+      return;
+    }
+
+    const payload = JSON.stringify({
+      ...notification.payload,
+      destination,
+    });
+
+    this.logger.debug(`Sending Slack DM notification: ${payload}`);
+
+    await sendSlackNotification(this.directWebhookUrl, payload);
+  }
+
+  async getEntity(entityRef: string): Promise<Entity> {
+    const { token } = await this.auth.getPluginRequestToken({
+      onBehalfOf: await this.auth.getOwnServiceCredentials(),
+      targetPluginId: 'catalog',
+    });
+
+    let entity = await this.cache.get<Entity>(entityRef);
+    if (!entity) {
+      const response = await this.catalogClient.getEntitiesByRefs(
```

Comment:
this sounds like it would benefit from some batching, put a `DataLoader` wrapper around this method I'd say